pkg/controller/argocd/role.go

```diff
@@ -0,0 +1,177 @@
+package argocd
+
+import (
+	"context"
+	"os"
+	"strings"
+
+	argoprojv1a1 "github.com/argoproj-labs/argocd-operator/pkg/apis/argoproj/v1alpha1"
+	"github.com/argoproj-labs/argocd-operator/pkg/common"
+	v1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func generateResourceName(argoComponentName string, cr *argoprojv1a1.ArgoCD) string {
+	return cr.Name + "-" + argoComponentName
+}
+
+// newRoleWithName creates a new ServiceAccount with the given name for the given ArgCD.
+func newRoleWithName(name string, cr *argoprojv1a1.ArgoCD) *v1.Role {
+	sa := newRole(name, cr)
+	sa.Name = name
+
+	lbls := sa.ObjectMeta.Labels
+	lbls[common.ArgoCDKeyName] = name
+	sa.ObjectMeta.Labels = lbls
+
+	return sa
+}
+
+// newRole returns a new ServiceAccount instance.
+func newRole(name string, cr *argoprojv1a1.ArgoCD) *v1.Role {
+	return &v1.Role{
+
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: cr.Namespace,
+			Labels:    labelsForCluster(cr),
+		},
+	}
+}
+
+// newClusterRoleWithName creates a new ClusterRole with the given name for the given ArgCD.
+func newClusterRoleWithName(name string, cr *argoprojv1a1.ArgoCD) *v1.ClusterRole {
+	sa := newClusterRole(name, cr)
+	sa.Name = generateResourceName(name, cr)
+
+	lbls := sa.ObjectMeta.Labels
+	lbls[common.ArgoCDKeyName] = name
+	sa.ObjectMeta.Labels = lbls
+
+	return sa
+}
+
+// newClusterRole returns a new ClusterRole instance.
+func newClusterRole(name string, cr *argoprojv1a1.ArgoCD) *v1.ClusterRole {
+	return &v1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   generateResourceName(name, cr),
+			Labels: map[string]string{},
+		},
+	}
+}
+
+func allowedNamespace(current string, configuredList string) bool {
+	isAllowedNamespace := false
+	if configuredList != "" {
+		if configuredList == "*" {
+			isAllowedNamespace = true
+		} else {
+			namespaceList := strings.Split(configuredList, ",")
+			for _, n := range namespaceList {
+				if n == current {
+					isAllowedNamespace = true
+				}
+			}
+		}
+	}
+	return isAllowedNamespace
+}
+
+// reconcileRoles will ensure that all ArgoCD Service Accounts are configured.
+func (r *ReconcileArgoCD) reconcileRoles(cr *argoprojv1a1.ArgoCD) error {
+	if err := r.reconcileRole(applicationController, policyRuleForApplicationController(), cr); err != nil {
+		return err
+	}
+
+	if err := r.reconcileRole(dexServer, policyRuleForDexServer(), cr); err != nil {
+		return err
+	}
+
+	if err := r.reconcileRole(server, policyRuleForServer(), cr); err != nil {
+		return err
+	}
+
+	if err := r.reconcileRole(redisHa, policyRuleForRedisHa(), cr); err != nil {
+		return err
+	}
+
+	rules := []v1.PolicyRule{}
+
+	if cr.Spec.ManagementScope.Cluster != nil && *cr.Spec.ManagementScope.Cluster {
+		if os.Getenv("ARGOCD_CLUSTER_CONFIG_ENABLED") == "true" &&
```

Comment:
Should the customer add this to argocd CR along with cr.Spec.ManagementScope.Cluster filed?

Reply:
Sorry, had missed this :(

If we add this to the CRD, users with low privileges would be able to spin up an ArgoCD for privilege escalation. 


Therefore, I suggest, supporting only one "ARGOCD_CLUSTER_CONFIG_NAMESPACES".

If set with the list of namespaces, use it. 
If unset, it implies that the admin has disabled cluster config argocd altogether/